java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -102,3 +104,88 @@ class MyBatisSqlOperationAnnotationMethod extends Method {
 class TypeParam extends Interface {
   TypeParam() { this.hasQualifiedName("org.apache.ibatis.annotations", "Param") }
 }
+
+private class MyBatisAbstractSQL extends RefType {
+  MyBatisAbstractSQL() { this.hasQualifiedName("org.apache.ibatis.jdbc", "AbstractSQL") }
+}
+
+private class MyBatisProvider extends RefType {
+  MyBatisProvider() {
+    this.hasQualifiedName("org.apache.ibatis.annotations",
+      ["Select", "Delete", "Insert", "Update"] + "Provider")
+  }
+}
+
+private class MyBatisAbstractSQLMethodNames extends string {
+  MyBatisAbstractSQLMethodNames() {
+    this in [
+        "SELECT", "OFFSET_ROWS", "FETCH_FIRST_ROWS_ONLY", "OFFSET", "LIMIT", "ORDER_BY", "HAVING",
+        "GROUP_BY", "WHERE", "OUTER_JOIN", "RIGHT_OUTER_JOIN", "LEFT_OUTER_JOIN", "INNER_JOIN",
+        "JOIN", "FROM", "DELETE_FROM", "SELECT_DISTINCT", "SELECT", "INTO_VALUES", "INTO_COLUMNS",
+        "VALUES", "INSERT_INTO", "SET", "UPDATE"
+      ]
+  }
+}
+
+/**
+ * A return statement of a method used in a MyBatis Provider.
+ *
+ * See
+ * - `MyBatisProvider`
+ * - https://mybatis.org/mybatis-3/apidocs/org/apache/ibatis/annotations/package-summary.html
+ */
+class MyBatisInjectionSink extends DataFlow::Node {
+  MyBatisInjectionSink() {
+    exists(Annotation a, Method m |
+      a.getType() instanceof MyBatisProvider and
+      m.getDeclaringType() = a.getValue(["type", "value"]).(TypeLiteral).getTypeName().getType() and
+      m.hasName(a.getValue("method").(StringLiteral).getValue()) and
+      this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+    )
+  }
+}
+
+private class MyBatisProviderStep extends TaintTracking::AdditionalValueStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(MethodAccess ma, Annotation a, Method providerMethod |
+      exists(int i |
+        ma.getArgument(i) = n1.asExpr() and
+        providerMethod.getParameter(i) = n2.asParameter()
+      )
+    |
+      a.getType() instanceof MyBatisProvider and
+      ma.getMethod().getAnAnnotation() = a and
+      providerMethod.getDeclaringType() =
+        a.getValue(["type", "value"]).(TypeLiteral).getTypeName().getType() and
+      providerMethod.hasName(a.getValue("method").(StringLiteral).getValue())
+    )
+  }
+}
+
+private class MyBatisAbstractSQLToStringStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row = ["org.apache.ibatis.jdbc;AbstractSQL;true;toString;;;Argument[-1];ReturnValue;taint"]
+  }
+}
+
+private class MyBatisAbstractSQLMethodsStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.apache.ibatis.jdbc;AbstractSQL;true;" + any(MyBatisAbstractSQLMethodNames m) +
+          ";;;Argument[0..1];ReturnValue;taint"
+      ]
```

Comment:
Is this too wide? I think adding each method would be a bit noisy, we would loose correlation with `MyBatisAbstractSQLMethodNames`, but CSV's `input` field would be more accurate.

Reply:
I'd favor being explicit here and wiring only the exact arguments we know propagate taint, even if it means repeating the method names.

If you wanted, you could always do that mapping in `MyBatisAbstractSQLMethodNames` (also, I think this should be singular) and create a predicate to obtain the tainted arguments depending on the method name, so something like:

```ql
private class MyBatisAbstractSQLMethodName extends string {
  string taintedArgs;

  MyBatisAbstractSQLMethodName() {
    // I'm making this up, I didn't check which method propagate taint from which arguments
    this in ["SELECT", "SELECT_DISTINCT"] and taintedArgs = "Argument[0]"
    or
    this in ["VALUES", "INSERT_INTO", "SET", "UPDATE"] and taintedArgs = "Argument[0..1]"
    // or ...
  }

  /** Gets the tainted arguments for this method name. */
  string getTaintedArgs() { result = taintedArgs }
}

```

and then:

```ql
private class MyBatisAbstractSQLMethodsStep extends SummaryModelCsv {
  override predicate row(string row) {
    exists(MyBatisAbstractSQLMethodName m |
      row =
        "org.apache.ibatis.jdbc;AbstractSQL;true;" + m + ";;;" + m.getTaintedArgs() +
          ";ReturnValue;taint"
    )
  }
}
```

This impacts readability though, so there isn't a perfect solution.

Reply:
It could be even more natural if `MyBatisAbstractSQLMethodName` becomes `MyBatisAbstractSQLMethod`:

```ql
private class MyBatisAbstractSQLMethod extends Method {
  string taintedArgs;

  MyBatisAbstractSQLMethod() {
    this.getDeclaringType().getSourceDeclaration() instanceof MyBatisAbstractSQL and
    (
      // I'm making this up, I didn't check which method propagate taint from which arguments
      this.hasName(["SELECT", "SELECT_DISTINCT"]) and taintedArgs = "Argument[0]"
      or
      this.hasName(["VALUES", "INSERT_INTO", "SET", "UPDATE"]) and taintedArgs = "Argument[0..1]"
      // or ...
    )
  }

  string getTaintedArgs() { result = taintedArgs }
}

// ...

private class MyBatisAbstractSQLMethodsStep extends SummaryModelCsv {
  override predicate row(string row) {
    exists(MyBatisAbstractSQLMethod m |
      row =
        "org.apache.ibatis.jdbc;AbstractSQL;true;" + m.getName() + ";;;" + m.getTaintedArgs() +
          ";ReturnValue;taint"
    )
  }
}

// ...

private class MyBatisAbstractSQLAnonymousClassStep extends TaintTracking::AdditionalTaintStep {
  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
    exists(MethodAccess ma, ClassInstanceExpr c |
      ma.getMethod() instanceof MyBatisAbstractSQLMethod and
      c.getAnonymousClass().getACallable() = ma.getCaller() and
      node1.asExpr() = ma and
      node2.asExpr() = c
    )
  }
}
```

Reply:
This is really cool!

Reply:
Done in https://github.com/github/codeql/pull/8345/commits/9aa440e5b6e9fd47d9ca12c749146be3a522deca